File: benchmark.jl
using BenchmarkTools
using InteractiveUtils

function add_vec1(A,B,n)
	C = Vector{Int32}(undef, n)
	for i in 1:n
		C[i] = A[i] + B[i]
	end
	return C
end

function add_vec2(A,B,C,n)
	for i in 1:n
		C[i] = A[i] + B[i]
	end
	return C
end

function add_vec3(A,B,C,n)
	for i in 1:2:n
		C[i] = A[i] + B[i]
		C[i+1] = A[i+1] + B[i+1]
	end
	return C
end



function benchmark1()
	n=Int32(10000)
	trials = Int32(1000)
	A=rand(Int32(1):Int32(1000),n)
	B=rand(Int32(1):Int32(1000),n)

	precompile(add_vec1,(Vector{Int32},Vector{Int32},Int32))
	precompile(add_vec2,(Vector{Int32},Vector{Int32},Vector{Int32},Int32))
	precompile(add_vec3,(Vector{Int32},Vector{Int32},Vector{Int32},Int32))


	println("Benchmark1: $(n*trials) additions")
	println("A,B are 32bit integer arrays of length $(length(A)), taking $(sizeof(A)) bytes each\n")

	println("add_vec1:\t")
	@time begin 
		for i in 1:trials
			add_vec1(A,B,n)
		end
	end
	println("add_vec2:\t")
	@time begin
		C = Vector{Int32}(undef, n)
		for i in 1:trials
			add_vec2(A,B,C,n)
		end
	end


	println("add_vec3:\t")
	@time begin
		C = Vector{Int32}(undef, n)
		for i in 1:trials
			add_vec3(A,B,C,n)
		end
	end

end
benchmark1()


# function benchmark2()


# function()
# end

# n=Int32(10000)
# A=rand(Int32(1):Int32(1000),n)
# B=rand(Int32(1):Int32(1000),n)
# @code_native add_vec1(A,B,n)


# n=Int32(10000)
# A=rand(Int32(1):Int32(1000),n)
# B=rand(Int32(1):Int32(1000),n)
# C = Vector{Int32}(undef, n)
# @code_native add_vec2(A,B,C,n)

File: benchmark.c
#include<time.h>
#include<stdio.h>
#include<stdlib.h>

int* vec_add1(int* A,int* B,int n){
	int *C = malloc(sizeof(int) * n);
	for(int i=0;i<n;i++){
		C[i] = A[i] + B[i];
	}
	return C;
}

int* vec_add2(int* A,int* B,int* C,int n){
	for(int i=0;i<n;i++){
		C[i] = A[i] + B[i];
	}
	return C;
}

int* vec_add3(int* A,int* B,int* C,int n){
	for(int i=0;i<n;i+=2){
		C[i] = A[i] + B[i];
		C[i+1] = A[i+1] + B[i+1];
	}
	return C;
}

void benchmark1(){
	struct timespec start, end;
	double elapsed;

	int n=10000;
	int trials = 1000;
	int* A = malloc(sizeof(int)*n);
	int* B = malloc(sizeof(int)*n);
	for(int i=0;i<n;i++){
		A[i] = rand() % 1000;
		B[i] = rand() % 1000;
	} 

	printf("Benchmark1: %d additions\n",n*trials);
	printf("A,B are 32bit integer arrays of length %d, taking %lu bytes each\n",n,sizeof(int)*n);



	//BENCHMARK CODE FOR vec_add1
	clock_gettime(CLOCK_MONOTONIC, &start);
	for(int i=0;i<trials;i++){
		int* C=vec_add1(A,B,n);
		free(C);
	}
	clock_gettime(CLOCK_MONOTONIC, &end);
	elapsed = ((end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / (1e9));
	printf("vec_add1 took %f seconds\n",elapsed);


	//BENCHMARK CODE FOR vec_add2
	clock_gettime(CLOCK_MONOTONIC, &start);
	int *C = malloc(sizeof(int) * n);
	for(int i=0;i<trials;i++){
		C=vec_add2(A,B,C,n);
	}
	free(C);
	clock_gettime(CLOCK_MONOTONIC, &end);
	elapsed = ((end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / (1e9));
	printf("vec_add2 took %f seconds\n",elapsed);

	
	//BENCHMARK code for vec_add3
	clock_gettime(CLOCK_MONOTONIC, &start);
	C = malloc(sizeof(int) * n);
	for(int i=0;i<trials;i++){
		C=vec_add3(A,B,C,n);
	}
	free(C);
	clock_gettime(CLOCK_MONOTONIC, &end);
	elapsed = ((end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / (1e9));
	printf("vec_add3 took %f seconds\n",elapsed);
	

	return;
}


int main(){
	// printf("%lu",sizeof)
	benchmark1();
	return 0;
}
